File: frontend/src/api/extraActivities.ts
import { useEffect, useState } from 'react';
import { readLearnerJson } from './learnerRead';
export interface ExtraActivity {
  activityId: string; title: string; status: string; submittedAt: string | null; month: string;
  answer: string; reflection: string; ksbs: string[]; hours: string;
  coachFeedback: string | null; reviewedBy: string | null; reviewedAt: string | null;
}
export function useExtraActivities(kind: string | undefined, learnerId: string | undefined, month?: string, enabled = true) {
  const [activities, setActivities] = useState<ExtraActivity[]>([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState('');
  const [version, setVersion] = useState(0);
  const [loadedFor, setLoadedFor] = useState('');
  const identity = `${kind}:${learnerId}:${month || ''}`;
  useEffect(() => {
    if (!enabled || !kind || !learnerId) return;
    const controller = new AbortController();
    setActivities([]); setLoading(true); setError('');
    const query = new URLSearchParams({ learnerKind: kind, learnerId, ...(month ? { month } : {}) });
    void readLearnerJson<{ activities: ExtraActivity[] }>(`/learner_api/reflection/extra-activities/?${query}`, { signal: controller.signal, revalidate: true })
      .then(data => { if (!Array.isArray(data.activities)) throw new Error('Invalid activities response'); if (!controller.signal.aborted) setActivities(data.activities); })
      .catch(() => { if (!controller.signal.aborted) setError('Could not load extra activities. Please retry.'); })
      .finally(() => { if (!controller.signal.aborted) { setLoadedFor(identity); setLoading(false); } });
    return () => controller.abort();
  }, [kind, learnerId, month, enabled, version, identity]);
  return { activities: loadedFor === identity ? activities : [], loading: enabled && (loading || loadedFor !== identity), error: loadedFor === identity ? error : '', refresh: () => setVersion(value => value + 1) };
}

File: backend/learner_api/extra_activities.py
"""Extra activities use the existing owned submission and coach-review records."""
import json
from uuid import UUID
from zoneinfo import ZoneInfo
from django.utils import timezone
from django.db import connections, DatabaseError
from django.http import JsonResponse
from login.permissions import learner_self_or_staff
from .monthly_assignment import assignment_checks, mapping, text

EXTRA_CHECK_KEYS = {'answer', 'learning', 'evidence', 'ksbs', 'planned', 'declarations', 'hours'}


def prepare_extra_activity(payload):
    identifier = text(payload.get('activityId'))
    if not identifier.startswith('extra:'):
        raise ValueError('A valid extra activity identifier is required.')
    try:
        UUID(identifier[6:])
    except ValueError:
        raise ValueError('A valid extra activity identifier is required.')
    current = timezone.now().astimezone(ZoneInfo('Europe/London'))
    return {**payload, 'monthlyAssignment': {**mapping(payload.get('monthlyAssignment')), 'month': current.strftime('%Y-%m'), 'timeEntries': mapping(payload.get('monthlyAssignment')).get('timeEntries', [])},
            'dateCompleted': current.date().isoformat(), 'plannedOtjh': ''}


def extra_activity_checks(payload, **dependencies):
    prepared = prepare_extra_activity(payload)
    checks = assignment_checks(prepared, meeting_booked=True, **dependencies)
    checks = [check for check in checks if check['key'] in EXTRA_CHECK_KEYS]
    return [{'key': 'title', 'label': 'Give your extra activity a title', 'passed': bool(text(payload.get('activityTitle')))}, *checks]


@learner_self_or_staff(query_param='learnerId')
def list_extra_activities(request):
    kind, learner_id = request.GET.get('learnerKind'), request.GET.get('learnerId')
    if kind not in ('commercial', 'apprenticeship') or not str(learner_id or '').isdigit():
        return JsonResponse({'error': 'A valid learner is required.'}, status=400)
    try:
        with connections['enrolment'].cursor() as cur:
            cur.execute('''SELECT activity_id, activity_title, status, submitted_at, full_submission,
                coach_feedback, reviewed_by, reviewed_at FROM "Learner".learning_reflection_submissions
                WHERE learner_kind=%s AND learner_id=%s AND activity_type='extra_activity'
                ORDER BY submitted_at DESC, id''', [kind, learner_id])
            rows = cur.fetchall()
    except DatabaseError:
        return JsonResponse({'error': 'Could not load extra activities. Please retry.'}, status=503)
    result = []
    for identifier, title, status, submitted, raw, feedback, reviewer, reviewed in rows:
        data = json.loads(raw) if isinstance(raw, str) else mapping(raw)
        month = submitted.astimezone(ZoneInfo('Europe/London')).strftime('%Y-%m') if submitted else ''
        if request.GET.get('month') and (status == 'draft' or month != request.GET['month']):
            continue
        result.append({'activityId': identifier, 'title': title, 'status': status,
                       'submittedAt': submitted.isoformat() if submitted else None, 'month': month,
                       'answer': data.get('assignmentAnswer', ''), 'reflection': data.get('whatYouLearned', ''),
                       'ksbs': data.get('ksbCodes', []), 'hours': data.get('actualTimeHours', ''),
                       'coachFeedback': feedback, 'reviewedBy': reviewer,
                       'reviewedAt': reviewed.isoformat() if reviewed else None})
    return JsonResponse({'activities': result})
